# Format ASV table and taxonomy

# Load libraries
library(tidyverse)
library(seqinr)


#### Load and format ASV data ####
# Load sequence table generated by dada2
# (Primer sequences were trimmed in Cutadapt)
load("output/sequence_tables/lp2017-2019watercolumn16sreseq2_seqtab_nochim.rda")

# Import taxonomic assignments generated by classifying sequences with TaxAss against
# FreshTrain15Jun2020silva138and silva_nr_v138
fasta <- read.fasta("output/taxass/otus.fasta", "DNA", as.string = TRUE, forceDNAtolower = FALSE)
fasta <- tibble(seqID = names(fasta),
                sequence = unlist(getSequence(fasta, as.string = TRUE)))

ncol(seqtab_nochim) == nrow(fasta)  # Should evaluate to TRUE

taxonomy <- read_csv("output/taxass/otus.98.80.80.taxonomy", col_names = TRUE)

taxonomy <- fasta %>%
  left_join(taxonomy, by = "seqID") %>%
  mutate(across(c(kingdom, phylum, class, order, lineage, clade, tribe), ~str_remove_all(.x, ".*__"))) %>%
  mutate(across(c(kingdom, phylum, class, order, lineage, clade, tribe), ~str_remove_all(.x, "\\(\\d.*"))) %>%
  select(-seqID) %>%
  mutate(across(c(kingdom, phylum, class, order, lineage, clade, tribe), ~case_when(grepl("unclassified", .x) ~ NA_character_,
                                                                                    TRUE ~ .x)))


#### Assign ASV codes ####
# Number of unique ASVs
(nasvs <- ncol(seqtab_nochim))

# Assign ASV codes to unique sequences
asv_codes <- tibble(sequence = colnames(seqtab_nochim)) %>%
  mutate(asv_code = paste0("ASV", str_pad(string = 1:nasvs, width = nchar(nasvs), side = "left", pad = "0")))

# Write ASVs to fasta file
# write.fasta(sequences = as.list(asv_codes$sequence),
#             names = asv_codes$asv_code,
#             file.out = "output/fasta/lp2017-2019watercolumn16sreseq2_all.fasta")


#### Combine sequence counts and taxonomy ####
# Convert sequence table to long format and join taxonomy
asv_melt <- seqtab_nochim %>%
  as_tibble(rownames = "sample_id") %>%
  pivot_longer(!sample_id, names_to = "sequence", values_to = "nseqs") %>%
  filter(nseqs > 0) %>%
  left_join(asv_codes, by = "sequence") %>%
  left_join(taxonomy, by = "sequence")

sum(asv_melt$nseqs) == sum(seqtab_nochim)  # Should evaluate to TRUE

# Write melted sequence data to file
# asv_melt %>%
#   write_tsv("output/melted/lp2017-2019watercolumn16sreseq2_melt_all.tsv", col_names = TRUE)
